aiopg/utils.py

```diff
@@ -122,3 +122,46 @@ async def __anext__(self) -> _TObj:
             finally:
                 self._obj = None
             raise
+
+
+class ClosableQueue:
+    """
+    Proxy object for an asyncio.Queue that is "closable"
+
+    When the ClosableQueue is closed, with an exception object as parameter,
+    subsequent or ongoing attempts to read from the queue will result in that
+    exception being result in that exception being raised.
+    """
+
+    def __init__(self, queue: asyncio.Queue):
+        self._queue = queue
+        self._close_exception = asyncio.Future()  # type: asyncio.Future[None]
+
+    def close(self, exception: Exception) -> None:
+        if not self._close_exception.done():
+            self._close_exception.set_exception(exception)
+
+    async def get(self) -> Any:
+        loop = get_running_loop()
+        get = loop.create_task(self._queue.get())
+
+        _, pending = await asyncio.wait(
+            [get, self._close_exception],
+            return_when=asyncio.FIRST_COMPLETED,
+        )
+
+        if get.done():
+            return get.result()
+        get.cancel()
+        self._close_exception.result()
+
+    def empty(self) -> bool:
+        return self._queue.empty()
+
+    def qsize(self) -> int:
+        return self._queue.qsize()
+
+    def get_nowait(self) -> Any:
+        if self._close_exception.done():
+            self._close_exception.result()
+        return self._queue.get_nowait()
```

Comment:
Behaviour of `get_nowait` differs from `get`: `get` allows to get all received notifications in case of the closed connection, but `get_nowait` throws exception right after a closure of the connection.

Reply:
BTW I don't know which behaviour is better 🤣 From the first glance a behaviour of `get` is more convenient.

Reply:
I'm not sure you are correct about `get()`. I hadn't thought about it.  What happens when there is an error but there are pending items in the queue?  :thinking: 

As far as I can tell, it's kind of undefined whether get() returns the next item or we raise the exception.  Because we are asking asyncio to wait for two futures, and they will both be ready at the same time, here:

```python
        _, pending = await asyncio.wait(
            [get, self._close_exception],
            return_when=asyncio.FIRST_COMPLETED,
        )
```
Does the `get` future become completed immediately, or does it need to be given a chance to execute by the main loop before becoming completed?  It's probably semi-random, which one is returned, to be honest.

If you think it matters, I can work some more on this to make sure that while there are items pending in the queue we keep consuming them before raising the exception.

Reply:
I've updated the code and docs so that both `get()` and `get_nowait()` keep consuming items until the queue is empty, and only check for the close exception when the queue becomes empty.